File: Data Structures & Algorithms/longest-substring-without-duplicates/submission-2.py
class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
        char_index = {}
        l = 0
        max_len = 0

        for r in range(len(s)):
            c = s[r]
            if c in char_index and char_index[c] >= l:
                l = char_index[c] + 1  # 移动左边界
            char_index[c] = r
            max_len = max(max_len, r - l + 1)

        return max_len

            

File: Data Structures & Algorithms/longest-repeating-substring-with-replacement/submission-2.java
class Solution {
    public int characterReplacement(String s, int k) {
        Map<Character, Integer> freq = new HashMap<>();
        int l = 0, maxFreq = 0, maxLen = 0;
        for(int r = 0; r<s.length();r++){
            char c = s.charAt(r);
            freq.put(c, freq.getOrDefault(c, 0)+1);
            maxFreq = Math.max(maxFreq, freq.get(c));
            while((r-l+1)-maxFreq>k){
                char left = s.charAt(l);
                freq.put(left, freq.get(left)-1);
                l++;
            }
            maxLen = Math.max(maxLen, r-l+1);
        }
        return maxLen;
}
}
